File: products-service/src/main/java/com/arielzarate/products/interfaces/errors/ErrorHandler.java
package com.arielzarate.products.interfaces.errors;


import com.arielzarate.products.interfaces.errors.exception.ApplicationErrorException;
import com.arielzarate.products.interfaces.errors.exception.ExternalClientException;
import com.arielzarate.products.interfaces.errors.model.ApplicationError;
import com.arielzarate.products.interfaces.errors.model.ClientError;
import jakarta.servlet.http.HttpServletRequest;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

@ControllerAdvice
@Slf4j
public class ErrorHandler {

    @ExceptionHandler(ExternalClientException.class)
    public ResponseEntity<ClientError> handleExternalClientException(ExternalClientException ex, HttpServletRequest request) {
        log.error("External client error: {}", ex.getMessage());
        ClientError error = new ClientError(
                "https://httpstatuses.com/" + ex.getClientError().status(),
                ex.getClientError().title(),
                ex.getClientError().status(),
                ex.getClientError().detail(),
                request.getRequestURI()
        );
        return ResponseEntity.status(error.status()).body(error);
    }

    @ExceptionHandler(ApplicationErrorException.class)
    public ResponseEntity<ClientError> handleApplicationErrorException(ApplicationErrorException ex, HttpServletRequest request) {
        //log.error("Application client error: {}", ex.getMessage());
        ApplicationError error = ex.getError();
        ClientError clientError = new ClientError(
                "https://httpstatuses.com/" + error.getStatus(),
                getTitleForStatus(error.getStatus()),
                error.getStatus(),
                error.getMessage(),
                request.getRequestURI()
        );
        return ResponseEntity.status(clientError.status()).body(clientError);
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ClientError> handleGenericException(Exception ex, HttpServletRequest request) {
        log.error("Unhandled exception: {}", ex.getMessage(), ex);
        ClientError error = new ClientError(
                "https://httpstatuses.com/500",
                "Internal Server Error",
                500,
                ex.getMessage(),
                request.getRequestURI()
        );
        return ResponseEntity.status(500).body(error);
    }

    private String getTitleForStatus(int status) {
        return switch (status) {
            case 400 -> "Bad Request";
            case 404 -> "Not Found";
            case 409 -> "Conflict";
            case 500 -> "Internal Server Error";
            default -> "Error";
        };
    }


}

File: shipping/src/main/kotlin/com/arielzarate/shipping/interfaces/error/ErrorHandler.kt
package com.arielzarate.shipping.interfaces.error

import com.arielzarate.shipping.interfaces.error.exception.ShipmentErrorException
import com.arielzarate.shipping.interfaces.error.exception.ExternalClientException
import com.arielzarate.shipping.interfaces.error.model.ShipmentError
import com.arielzarate.shipping.interfaces.error.model.ClientError
import jakarta.servlet.http.HttpServletRequest
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler

@ControllerAdvice
class ErrorHandler {

    private val log = LoggerFactory.getLogger(ErrorHandler::class.java)

    @ExceptionHandler(ExternalClientException::class)
    fun handleExternalClientException(ex: ExternalClientException, request: HttpServletRequest): ResponseEntity<ClientError> {
        log.error("External client error: ${ex.message}")
        val error = ClientError(
            type = "https://httpstatuses.com/${ex.clientError.status}",
            title = ex.clientError.title,
            status = ex.clientError.status,
            detail = ex.clientError.detail,
            instance = request.requestURI
        )
        return ResponseEntity.status(error.status).body(error)
    }

    @ExceptionHandler(ShipmentErrorException::class)
    fun handleApplicationErrorException(ex: ShipmentErrorException, request: HttpServletRequest): ResponseEntity<ClientError> {
        val appError: ShipmentError = ex.error
        val clientError = ClientError(
            type = "https://httpstatuses.com/${appError.status}",
            title = getTitleForStatus(appError.status),
            status = appError.status,
            detail = appError.message,
            instance = request.requestURI
        )
        return ResponseEntity.status(clientError.status).body(clientError)
    }

    @ExceptionHandler(Exception::class)
    fun handleGenericException(ex: Exception, request: HttpServletRequest): ResponseEntity<ClientError> {
        log.error("Unhandled exception: ${ex.message}", ex)
        val error = ClientError(
            type = "https://httpstatuses.com/500",
            title = "Internal Server Error",
            status = 500,
            detail = ex.message ?: "No detail",
            instance = request.requestURI
        )
        return ResponseEntity.status(500).body(error)
    }

    private fun getTitleForStatus(status: Int): String = when (status) {
        400 -> "Bad Request"
        404 -> "Not Found"
        409 -> "Conflict"
        500 -> "Internal Server Error"
        else -> "Error"
    }
}
